src/QsCompiler/Compiler/CompilationLoader.cs

```diff
@@ -1206,5 +1206,26 @@ public static string GeneratedFile(string fileId, string outputFolder, string fi
             File.WriteAllText(targetFile, content);
             return targetFile;
         }
+
+        /// <summary>
+        /// Writes the given contents to the given file name at the given output folder.
+        /// Throws the corresponding exception if any of the path operations fails or if the writing fails.
+        /// </summary>
+        public static void WriteFile(string fileName, string outputFolder, string content)

```

Comment:
Why is this separate method needed opposed to using GeneratedFile?

Reply:
GeneratedFile only works on files that have a matching Q# source file. We need something more, like this, to generate a separate file for the C# Main.

Reply:
Do you mean because of the use of TryGetUri? No Q# file needs to exist - the only requirement is that the format of the path matches what the compiler expects (take a look at TryGetUri - I believe it was a full absolute path).

Reply:
The `TryGetUri` will not succeed (it will return false with the out var as null) if the given file path does not point to an existing folder, i.e. an existing Q# file. This is probably a good check when generating files that are based on existing Q# files, like what is done in `GeneratedFile`, but is not a valid check when generating the Main file, which is not tied to any one specific Q# file.

Reply:
There is also no reason to get a Uri to any existing file when generating the Main file. At least as far as I can tell.

Reply:
Hmm, maybe I'm not right about `TryGetUri` failing if the file doesn't exist. But then what would I provide for arguments to the `GeneratedFile` method to produce the `EntryPoint.g.Main.cs` file?

Reply:
You just need to full path for the file (opposed to just a local path). the fileId is then the full path (e.g. C:/SomePath/EntryPoint.g.Mail.cs), the output folder would be e.g. C:/SomePath/ (don't count on my slashes being right), the file ending would be .cs (or you could also give a fileId C:/SomePath/EntryPoint.qs and then a file ending of .g.Main.cs). 

To make sure you have the correct slashes, the easiest is to create an Uri and then use CompilationUnitManager.TryGetFileId.

Reply:
I see, thanks! So the end result on the runtime repo will look like:
```
let mainSourceFile = "./EntryPoint" |> Path.GetFullPath |> Uri |> CompilationUnitManager.GetFileId
CompilationLoader.GeneratedFile(mainSourceFile, dir, ".g.Main.cs", content) |> ignore
```
This will remove the need for the `WriteFile` method on the `CompilationLoader`. Nice!

Reply:
Sounds good. I think you are missing the output folder that is configurable for the C# generation here somewhere (instead of "./EntryPoint" is should probabably be something like Path.Combine(outputfolder, "EntryPoint")).

Reply:
The file id is just used to get the file name, not the path. In the `GeneratedFile` method it will use the output folder (passed in here as `dir`) and append the file name from the file id, and append the given file extension. So it doesn't matter what path we use in the file id, so long as it ends with the name we want for the resulting file, in this case "EntryPoint". It does have to be a valid, absolute path though.

Reply:
The GeneratedFile method preserves the folder structure in the output folder. I checked and the structure is preserved relative to the current directory, which means what you are saying will work, though it feels a little bit dangerous to me to give a path that you know is not "correct". Just giving this as input to consider.

Reply:
Can you give me an example of what you mean when you say 'The GeneratedFile method preserves the folder structure in the output folder.'? If what I think you are saying is correct, then https://github.com/microsoft/qsharp-runtime/issues/357 wouldn't be an issue.

Reply:
@bettinaheim Oh I see, that issue is about files not under the current Q# project's folder structure. Okay, I now really think that the task of fixing the file name collisions has to be separated out into some separate work. There just keeps being more and more nuances and complexity added onto this, and it really doesn't have anything to do with Multiple Entry Points. I'm going to remove the enumeration in the name logic and try to get the Multiple Entry Points feature in without it.

Reply:
Okay so given that there will be a folder structure under the output folder, will feeding the "./EntryPoint" into the GeneratedFile method always put the EntryPoint.g.Main.cs directly in the output folder (not a subfolder), or is it unsafe enough to were there is a possibility where it could go somewhere else? And would saying `Path.Combine(outputfolder, "EntryPoint")` fix that issue to make sure it is always directly in the output folder?

Reply:
I'm moving the conversation on the `"./EntryPoint"` vs `Path.Combine(outputFolder, "EntryPoint")` to the runtime PR since it is more relevant there. Here is the link: https://github.com/microsoft/qsharp-runtime/pull/528#discussion_r598423034

Reply:
Yes, that definitively applied more to the runtime PR than this one, sorry.